File: android-native/core/data/src/main/kotlin/today/mindlog/todo/core/data/OAuthCallbackBus.kt
package today.mindlog.todo.core.data

import android.net.Uri
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import javax.inject.Inject
import javax.inject.Singleton

/** What the browser can hand back on the deep link. */
sealed interface OAuthCallback {
    data class Tokens(val accessToken: String, val refreshToken: String) : OAuthCallback

    /**
     * mindlog id returned a profile with no email address. The server issues a
     * short-lived pending token and the app has to collect one.
     */
    data class NeedsEmail(val pendingToken: String) : OAuthCallback

    data class Failed(val reason: String) : OAuthCallback
}

/**
 * Carries the OAuth return leg from the activity to whichever screen is
 * waiting.
 *
 * The deep link lands on `MainActivity.onNewIntent`, but the outcome belongs to
 * the sign-in screen — and one of the outcomes needs a text field. A shared
 * flow keeps the activity from having to know that.
 */
@Singleton
class OAuthCallbackBus @Inject constructor() {

    private val _callbacks = MutableSharedFlow<OAuthCallback>(extraBufferCapacity = 1)
    val callbacks: SharedFlow<OAuthCallback> = _callbacks.asSharedFlow()

    /**
     * Parses `…://auth/callback#access_token=…&refresh_token=…&expires_in=…`,
     * or `#mindlog_id_pending=…`.
     *
     * The tokens travel in the **fragment**, not the query, so they never reach
     * a server log or a Referer header. The web client reads the same shape.
     */
    fun onDeepLink(uri: Uri): Boolean {
        val fragment = uri.fragment ?: return false
        val params = fragment.split('&')
            .mapNotNull { part ->
                val i = part.indexOf('=')
                if (i <= 0) null else Uri.decode(part.take(i)) to Uri.decode(part.substring(i + 1))
            }
            .toMap()

        val callback = when {
            params["access_token"] != null && params["refresh_token"] != null ->
                OAuthCallback.Tokens(params.getValue("access_token"), params.getValue("refresh_token"))

            params["mindlog_id_pending"] != null ->
                OAuthCallback.NeedsEmail(params.getValue("mindlog_id_pending"))

            params["error"] != null -> OAuthCallback.Failed(params.getValue("error"))

            else -> return false
        }

        return _callbacks.tryEmit(callback)
    }
}

File: ios-native/Sources/CoreData/OAuthCallbackBus.swift
import Foundation

/// What the browser can hand back on the return leg.
public enum OAuthCallback: Sendable, Equatable {
    case tokens(accessToken: String, refreshToken: String)

    /// mindlog id returned a profile with no email address. The server issues a
    /// short-lived pending token and the app has to collect one.
    case needsEmail(pendingToken: String)

    case failed(reason: String)
}

/// Carries the OAuth return leg to whichever screen is waiting.
///
/// On Android the deep link lands on `MainActivity.onNewIntent` and the outcome
/// belongs to the sign-in screen, so a shared flow sits between them. Here
/// `ASWebAuthenticationSession` usually hands the callback URL straight back to
/// the caller — but not always: the user can be bounced through Safari and
/// return through the app's URL scheme instead. Both paths end at ``handle(_:)``
/// so the sign-in screen has one thing to listen to.
@MainActor
public final class OAuthCallbackBus {

    public let callbacks: AsyncStream<OAuthCallback>
    private let continuation: AsyncStream<OAuthCallback>.Continuation

    public init() {
        (callbacks, continuation) = AsyncStream<OAuthCallback>.makeStream(
            bufferingPolicy: .bufferingNewest(1)
        )
    }

    /// Parses `…://auth/callback#access_token=…&refresh_token=…&expires_in=…`,
    /// or `#mindlog_id_pending=…`.
    ///
    /// The tokens travel in the **fragment**, not the query, so they never reach
    /// a server log or a Referer header. The web and Android clients read the
    /// same shape.
    ///
    /// - Returns: whether the URL was one of ours.
    @discardableResult
    public func handle(_ url: URL) -> Bool {
        guard let fragment = URLComponents(url: url, resolvingAgainstBaseURL: false)?
            .percentEncodedFragment
        else { return false }

        let params = Self.parse(fragment: fragment)

        let callback: OAuthCallback
        if let accessToken = params["access_token"], let refreshToken = params["refresh_token"] {
            callback = .tokens(accessToken: accessToken, refreshToken: refreshToken)
        } else if let pending = params["mindlog_id_pending"] {
            callback = .needsEmail(pendingToken: pending)
        } else if let error = params["error"] {
            callback = .failed(reason: error)
        } else {
            return false
        }

        continuation.yield(callback)
        return true
    }

    /// Decoded pair by pair rather than through `URLComponents(query:)`: the
    /// fragment is not a query, and a token containing `+` would come back with
    /// a space in it if it were treated as one.
    private static func parse(fragment: String) -> [String: String] {
        fragment.split(separator: "&").reduce(into: [:]) { params, part in
            guard let separator = part.firstIndex(of: "="),
                  separator != part.startIndex
            else { return }
            let name = String(part[part.startIndex ..< separator])
            let value = String(part[part.index(after: separator)...])
            guard let name = name.removingPercentEncoding,
                  let value = value.removingPercentEncoding
            else { return }
            params[name] = value
        }
    }
}
